File: frontend/src/types/listing.ts
export interface CatalogCrop {
  id: string;
  slug: string;
  commonName: string;
  scientificName: string | null;
  category: string | null;
  description: string | null;
}

export interface CatalogVariety {
  id: string;
  cropId: string;
  slug: string;
  name: string;
  description: string | null;
}

export interface GrowerCropItem {
  id: string;
  userId: string;
  canonicalId: string | null;
  cropName: string;
  varietyId: string | null;
  status: string;
  visibility: string;
  surplusEnabled: boolean;
  nickname: string | null;
  defaultUnit: string | null;
  notes: string | null;
  createdAt: string;
  updatedAt: string;
}

export interface Listing {
  id: string;
  userId: string;
  growerCropId: string | null;
  cropId: string;
  varietyId: string | null;
  title: string;
  unit: string;
  quantityTotal: string;
  quantityRemaining: string;
  availableStart: string;
  availableEnd: string;
  status: string;
  pickupLocationText: string | null;
  pickupAddress: string | null;
  pickupDisclosurePolicy: string;
  pickupNotes: string | null;
  contactPref: string;
  geoKey: string | null;
  lat: number;
  lng: number;
  createdAt: string;
}

export interface ListMyListingsResponse {
  items: Listing[];
  limit: number;
  offset: number;
  hasMore: boolean;
  nextOffset: number | null;
}

export interface DiscoverListingsResponse {
  items: Listing[];
  limit: number;
  offset: number;
  hasMore: boolean;
  nextOffset: number | null;
}

export interface UpsertListingRequest {
  title: string;
  cropId: string;
  varietyId?: string;
  quantityTotal: number;
  unit: string;
  availableStart: string;
  availableEnd: string;
  pickupLocationText?: string;
  pickupAddress?: string;
  pickupDisclosurePolicy?: 'immediate' | 'after_confirmed' | 'after_accepted';
  pickupNotes?: string;
  contactPref?: 'app_message' | 'phone' | 'knock';
  lat: number;
  lng: number;
  status?: 'active' | 'pending' | 'claimed' | 'expired' | 'completed';
}

File: services/grn-api/src/api/models/listing.rs
use serde::{Deserialize, Serialize};

#[derive(Debug, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ListingItem {
    pub id: String,
    pub user_id: String,
    pub grower_crop_id: Option<String>,
    pub crop_id: Option<String>,
    pub variety_id: Option<String>,
    pub title: Option<String>,
    pub unit: Option<String>,
    pub quantity_total: Option<String>,
    pub quantity_remaining: Option<String>,
    pub available_start: Option<String>,
    pub available_end: Option<String>,
    pub status: String,
    pub pickup_location_text: Option<String>,
    pub pickup_address: Option<String>,
    pub effective_pickup_address: Option<String>,
    pub pickup_disclosure_policy: String,
    pub pickup_notes: Option<String>,
    pub contact_pref: String,
    pub geo_key: Option<String>,
    pub lat: Option<f64>,
    pub lng: Option<f64>,
    pub created_at: String,
}

#[derive(Debug, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ListMyListingsResponse {
    pub items: Vec<ListingItem>,
    pub limit: i64,
    pub offset: i64,
    pub has_more: bool,
    pub next_offset: Option<i64>,
}

#[derive(Debug, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct DiscoverListingsResponse {
    pub items: Vec<ListingItem>,
    pub limit: i64,
    pub offset: i64,
    pub has_more: bool,
    pub next_offset: Option<i64>,
}
